geth/jail/jail.go

```diff
@@ -165,61 +168,70 @@ func (jail *Jail) Send(call otto.FunctionCall) (response otto.Value) {
 
 	if rawReq[0] == '[' {
 		batch = true
-		json.Unmarshal(rawReq, &reqs)
+		err = json.Unmarshal(rawReq, &reqs)
 	} else {
 		batch = false
 		reqs = make([]common.RPCCall, 1)
-		json.Unmarshal(rawReq, &reqs[0])
+		err = json.Unmarshal(rawReq, &reqs[0])
+	}
+	if err != nil {
+		throwJSException(fmt.Errorf("can't unmarshal %v (batch=%v): %s", string(rawReq), batch, err))
 	}
 
-	resps, _ := call.Otto.Object("new Array()")
+	resps, err := vm.Call("new Array", nil)
+	if err != nil {
+		throwJSException(fmt.Errorf("can't create Array: %s", err))
+	}
 
 	// Execute the requests.
 	for _, req := range reqs {
 		log.Info("execute request", "method", req.Method)
-		res, err := jail.policy.Execute(req, call)
+		res, err := jail.policy.Execute(req, vm)
 		if err != nil {
 			log.Info("request errored", "error", err.Error())
 			switch err.(type) {
 			case common.StopRPCCallError:
-				return newErrorResponse(call.Otto, -32603, err.Error(), nil)
+				return newErrorResponseOtto(vm, err.Error(), nil)
 			default:
-				res = newErrorResponse(call.Otto, -32603, err.Error(), &req.ID).Object()
+				res = newErrorResponse(err.Error(), &req.ID)
 			}
 		}
 
-		resps.Call("push", res)
+		_, err = resps.Object().Call("push", res)
+		if err != nil {
+			throwJSException(fmt.Errorf("can't push result: %s", err))
+		}
 	}
 
 	// Return the responses either to the callback (if supplied)
 	// or directly as the return value.
 	if batch {
-		response = resps.Value()
-	} else {
-		response, _ = resps.Get("0")
+		return resps
 	}
-
-	if fn := call.Argument(1); fn.Class() == "Function" {
-		fn.Call(otto.NullValue(), otto.NullValue(), response)
-		return otto.UndefinedValue()
+	v, err := resps.Object().Get("0")
+	if err != nil {
+		throwJSException(err)
 	}
-
-	return response
+	return v
 }
 
-func newErrorResponse(vm *otto.Otto, code int, msg string, id interface{}) otto.Value {
+func newErrorResponse(msg string, id interface{}) map[string]interface{} {
 	// Bundle the error into a JSON RPC call response
-	m := map[string]interface{}{
+	return map[string]interface{}{
 		"jsonrpc": "2.0",
 		"id":      id,
 		"error": map[string]interface{}{
-			"code":    code,
+			"code":    -32603, // Internal JSON-RPC Error, see http://www.jsonrpc.org/specification#error_object
 			"message": msg,
 		},
 	}
-	res, _ := json.Marshal(m)
-	val, _ := vm.Run("(" + string(res) + ")")
-	return val
+}
+
+func newErrorResponseOtto(vm *vm.VM, msg string, id interface{}) otto.Value {
+	// TODO(tiabc): Handle errors.
+	errResp, _ := json.Marshal(newErrorResponse(msg, id))
+	errRespVal, _ := vm.Run("(" + string(errResp) + ")")
```

Comment:
This looks a bit cryptic. Why do we call JS to form error?

Reply:
Shouldnt we assign it into a variable in the script source to be runned?

Reply:
@divan this is legacy. But it seems that's to get `otto.Value` out of an arbitrary object.
@influx6 nope, you can run it and get the result into `errRespVal`.

Reply:
i see. Understood.